neo4j/internal/bolt/connect.go

```diff
@@ -115,3 +134,181 @@ func Connect(ctx context.Context,
 	}
 	return boltConn, nil
 }
+
+// performManifestNegotiation handles the manifest-style handshake.
+// Returns the negotiated protocol's major and minor version.
+func performManifestNegotiation(
+	ctx context.Context,
+	conn io.ReadWriteCloser,
+	serverName string,
+	errorListener ConnectionErrorListener,
+	boltLogger log.BoltLogger,
+	response []byte,
+) (byte, byte, error) {
+	reader := racing.NewRacingReader(conn)
+
+	// Read the protocol offerings.
+	count, supported, err := readProtocolOfferings(ctx, reader, serverName, errorListener)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	// Read the capability mask.
+	_, capBytes, err := readCapabilityMask(ctx, reader, serverName, errorListener)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	// Log the complete server handshake message.
+	logManifestHandshake(boltLogger, response, count, supported, capBytes)
+
+	// Select an acceptable protocol version.
+	chosen, err := selectProtocol(supported, errorListener, serverName)
+	if err != nil {
+		invalidHandshake := []byte{0x00, 0x00, 0x00, 0x00, 0x00} // 4 bytes for version + 1 byte for capabilities.
+		if _, err := conn.Write(invalidHandshake); err != nil {
+			errorListener.OnDialError(ctx, serverName, err)
+		}
+		return 0, 0, err
+	}
+
+	// Send the handshake confirmation.
+	if err = sendHandshakeConfirmation(ctx, conn, boltLogger, errorListener, serverName, chosen, capBytes); err != nil {
+		return 0, 0, err
+	}
+
+	return chosen.major, chosen.minor, nil
+}
+
+// readProtocolOfferings reads the number of protocol offerings and returns the count and
+// a slice of supported protocol versions.
+func readProtocolOfferings(ctx context.Context, r racing.RacingReader, serverName string, errorListener ConnectionErrorListener) (uint64, []protocolVersion, error) {
+	count, err := readVarInt(ctx, r)
+	if err != nil {
+		errorListener.OnDialError(ctx, serverName, err)
+		return 0, nil, fmt.Errorf("failed to read manifest protocol count: %w", err)
+	}
+	supported := make([]protocolVersion, count)
```

Comment:
```suggestion
	supported := make([]protocolVersion, 0, count)
```
No need to initialize all protocolVersion structs just to then overwrite them further down.
If accepted, also change `supported[i] = protocolVersion{...}` to use `append` instead.